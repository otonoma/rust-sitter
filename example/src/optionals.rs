#[allow(dead_code)]
mod grammar {
    use rust_sitter::Spanned;
    use rust_sitter::Rule;

    #[derive(Debug, Rule)]
    #[language]
    pub struct Language {
        #[leaf(re(r"\d+"))]
        // Not necessary, done automatically.
        // #[rust_sitter::with(|v| v.parse().unwrap())]
        v: Option<i32>,
        #[leaf("_")]
        _s: (),
        t: Spanned<Option<Number>>,
        #[leaf(".")]
        _d: Option<()>,
    }

    #[derive(Debug, Rule)]
    pub struct Number {
        #[leaf(re(r"\d+"))]
        // TODO: We are replacing this entirely with a different defintion.
        // #[with(|v| v.parse().unwrap())]
        v: i32,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rust_sitter::Language;

    #[test]
    fn optional_grammar() {
        insta::assert_debug_snapshot!(grammar::Language::parse("_"));
        insta::assert_debug_snapshot!(grammar::Language::parse("_."));
        insta::assert_debug_snapshot!(grammar::Language::parse("1_"));
        insta::assert_debug_snapshot!(grammar::Language::parse("1_."));
        insta::assert_debug_snapshot!(grammar::Language::parse("1_2"));
        insta::assert_debug_snapshot!(grammar::Language::parse("1_2."));
        insta::assert_debug_snapshot!(grammar::Language::parse("_2"));
        insta::assert_debug_snapshot!(grammar::Language::parse("_2."));
    }
}
