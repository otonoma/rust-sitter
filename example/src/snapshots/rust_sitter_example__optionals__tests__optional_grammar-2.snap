---
source: example/src/optionals.rs
expression: "grammar::Language::parse(\"_.\")"
---
ParseResult {
    result: Some(
        Language {
            v: None,
            _s: (),
            t: Spanned {
                value: None,
                position: Position {
                    bytes: 0..0,
                    start: Point {
                        line: 1,
                        column: 1,
                    },
                    end: Point {
                        line: 1,
                        column: 1,
                    },
                },
            },
            _d: Some(
                (),
            ),
        },
    ),
    errors: [],
}
